pandas/core/groupby/indexing.py

```diff
@@ -0,0 +1,262 @@
+
+        - An integer valued iterable, e.g. ``range(2, 4)``.
+        - A comma separated list of integers and slices, e.g. ``5``, ``2, 4``, ``2:4``.
+
+        Note: the slice step cannot be negative.
+
+        The output format is the same as GroupBy.head and GroupBy.tail, namely a subset
+        of the original DataFrame or Series with the index and order preserved.
+
+        The effect of ``grouped.rows[i:j]`` is similar to
+
+            ``grouped.apply(lambda x: x.iloc[i:j])``
+
+        but much faster and preserving the original index and order.
+
+        The behaviour is different from GroupBy.nth:
+
+        - Input to rows can include one or more slices whereas nth just handles
+          a list of indexes.
+        - Output from rows is in the same order as the original grouped DataFrame
+          or Series.
+        - Output from rows has the same index columns as the original grouped DataFrame
+          or Series. (nth behaves like an aggregator and removes the non-grouped
+          indexes.)
+        - GroupBy.rows can  define a slice relative to the last row of each group.
+        - GroupBy.rows is faster than nth.
+        - GroupBy.rows does not handle dropna.
+
+        An important use case for GroupBy.rows is a multi-indexed DataFrame with a
+        large primary index (Date, say) and a secondary index sorted to a different
+        order for each Date.
+        To reduce the DataFrame to a middle slice of each Date:
+
+            ``df.groupby("Date").rows[5:-5]``
+
+        This returns a subset of df containing just the middle rows for each Date
+        and with its original order and indexing preserved.
+
+        To reduce the DataFrame to the remaining rows:
+
+            ``df.groupby("Date").rows[:5, -5:]``
+
+        Returns
+        -------
+        Series
+            The filtered subset of the original grouped Series.
+        DataFrame
+            The filtered subset of the original grouped DataFrame.
+
+        See Also
+        --------
+        DataFrame.iloc : Purely integer-location based indexing for selection by
+            position.
+        GroupBy.head : Return first n rows of each group.
+        GroupBy.tail : Return last n rows of each group.
+        GroupBy.nth : Take the nth row from each group if n is an int, or a
+            subset of rows, if n is a list of ints.
+
+        Examples
+        --------
+            >>> df = pd.DataFrame([["a", 1], ["a", 2], ["a", 3], ["b", 4], ["b", 5]],
+            ...                   columns=["A", "B"])
+            >>> df.groupby("A", as_index=False).rows[1:2]
+               A  B
+            1  a  2
+            4  b  5
+
+            >>> df.groupby("A", as_index=False).rows[1, -1]
+               A  B
+            1  a  2
+            2  a  3
+            4  b  5
+        """
+        return _rowsGroupByIndexer(self)
+
+
+@doc(GroupByIndexingMixin.rows)
+class _rowsGroupByIndexer:
+    def __init__(self, grouped):
```

Comment:
type in & out args (e.g. this is FrameOrSeries)

Reply:
Painful, but done!